aip/general/0193.md

```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+dynamic information as context. Each key within `metadata` **must** be at most
+64 characters long, and conform to the regular expression `[a-z][a-zA-Z0-9-_]+`.
 
-> The Book, "The Great Gatsby", is unavailable at the Library, "Garfield
-> East".  It is expected to be available again on 2199-05-13.
+Any request-specific information which contributes to the `Status.message` or
+`LocalizedMessage.message` messages **must** be represented within `metadata`.
+This practice is critical so that machine actors do not need to parse error
+messages to extract information.
 
-The preceding error message is made actionable by the context, both
-originating from the request, the title of the Book, the name of the
-Library, and by the information that is known only by the service, that
-is, the expected return date of the Book.
+For example consider the following message:
 
-All dynamic variables found in error messages **must** also be present
-in the `map<string, string>`, `ErrorInfo.metadata` (found on the
-*required* `ErrorInfo`). For example, the `metadata` map for the sample
-error message above will include *at least* the following key/value
-pairs:
+> An &lt;e2-medium&gt; VM instance with &lt;local-ssd=3,nvidia-t4=2&gt; is currently unavailable
+> in the &lt;us-east1-a&gt; zone. Consider trying your request in the &lt;us-central1-f,us-central1-c&gt;
+> zone(s), which currently has/have capacity to accommodate your request. Alternatively,
+> you can try your request again with a different VM hardware configuration
+> or at a later time. For more information, see the troubleshooting documentation.
 
-```
-bookTitle: "The Great Gatsby"
-library: "Garfield East"
-expectedReturnDate: "2199-05-13"
-```
+The `ErrorInfo.metadata` map for the same error could be:
 
-The following example shows an additional example of a metadata map for
-information about virtual machines:
+- "zone": "us-east1-a",
+- "vmType": "e2-medium",
+- "attachment": "local-ssd=3,nvidia-t4=2",
+- "zonesWithCapacity": "us-central1-f,us-central1-c"
 
-```
-vmType: "<VM-TYPE>",
-attachment: "<ATTACHMENT>"
-zone: "<ZONE>"
-```
+Additional contextual information that does not appear in an error message
+**may** also be included in `metadata` to allow programmatic use by the client.
+
+The metadata included for any given (reason,domain) pair can evolve over time:
 
-Dynamic variables that do not appear in an error message **may** also be
-included in `metadata` to provide additional information to the client
-to be used programmatically.
+- New keys **may** be included
+- All keys that have been included **must** continue to be included (but may have empty values)
 
-Once present in `metadata`, keys **must** continue to be included in the
-map for the error payload to be backwards compatible, even if the value
-for a particular key is empty. Keys **must** be expressed as lower
-camel-case.
+In other words, once a user has observed a given key for a (reason, domain) pair they should be
```

Comment:
```suggestion
In other words, once a user has observed a given key for a (reason, domain) pair they **must** be
```
stronger guidance? If not, perhaps just bold the `should`...or change to something else if you want to just make an assertion e.g. `the need to be able to rely...`

Reply:
Rewritten - I tend to think that every "must" and "should" ought to be aimed at the target audience of the AIP - usually the service owner, rather than the user. (Client library AIPs are slightly different.) Each should/must ought to be effectively an instruction - whereas saying "the user must be able to do something" isn't really an instruction to the subject of the verb... see what you think of this rewrite :)

Reply:
LGTM.